src/centralserver/registration-service/src/main/java/org/niis/xroad/centralserver/registrationservice/controller/RegistrationRequestController.java

```diff
@@ -0,0 +1,103 @@
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.input.BoundedInputStream;
+import org.niis.xroad.centralserver.registrationservice.request.ManagementRequestHandler;
+import org.niis.xroad.centralserver.registrationservice.request.ManagementRequestParser;
+import org.niis.xroad.centralserver.registrationservice.request.ManagementRequestUtil;
+import org.niis.xroad.centralserver.registrationservice.service.AdminApiService;
+import org.springframework.http.CacheControl;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestHeader;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.io.InputStream;
+
+/**
+ * Handles security server authentication certificate registration requests from
+ * security servers.
+ *
+ * Security note: This API is exposed publicly (security servers not yet part of this X-Road instance need to be
+ * able to send the registration request).
+ */
+@RestController
+@Slf4j
+@RequiredArgsConstructor
+public class RegistrationRequestController {
+
+    public static final int MAX_REQUEST_SIZE = 100_000;
+    private final AdminApiService adminApiService;
+
+    @ResponseBody
+    @PostMapping(path = "/managementservice",
+            produces = {MediaType.TEXT_XML_VALUE},
+            consumes = {MediaType.MULTIPART_RELATED_VALUE})
+    public ResponseEntity<String> register(@RequestHeader("Content-Type") String contentType, InputStream body) {
+
+        try {
+            var bos = new BoundedInputStream(body, MAX_REQUEST_SIZE);
+            var message = ManagementRequestHandler.readRequest(contentType, bos);
+            var authRequest = ManagementRequestParser.parseAuthCertRegRequest(message);
+
+            var requestId = adminApiService.addRegistrationRequest(
+                    authRequest.getServer(),
+                    authRequest.getAddress(),
+                    authRequest.getAuthCert());
+
+            return ResponseEntity
+                    .ok()
+                    .header("Pragma", "no-cache").header("Expires", "0")
+                    .cacheControl(CacheControl.noStore())
+                    .body(ManagementRequestUtil.toResponse(message, requestId).getXml());
+        } catch (Exception e) {
+            var ex = ErrorCodes.translateException(e);
+            if (log.isDebugEnabled() || !(e instanceof CodedException)) {
+                log.error("Registration failed [{}]", ex.getFaultDetail(), ex);
+            } else {
+                var cause = (ex.getCause() == null) ? "" : ex.getCause().toString();
+                log.error("Registration failed [{}]: {}: {}", ex.getFaultDetail(), ex.getMessage(), cause);
+            }
+            return ResponseEntity
+                    .internalServerError()
```

Comment:
This is due to the "SOAP endpoint" nature of the the interface:
> In case of a SOAP error while processing the request, the SOAP HTTP server MUST issue an HTTP 500 "Internal Server Error" response  (https://www.w3.org/TR/2000/NOTE-SOAP-20000508/)